pkg/sidecar/inject.go

```diff
@@ -0,0 +1,238 @@
+		}
+	}
+
+	port, err := findPort(pod, ADAPTER_PORT)
+	if err != nil {
+		errs = errs.Also(apis.ErrGeneric("No free port: " + err.Error()).ViaField("spec.containers[i].ports"))
+	}
+
+	ceSrc, labelerr := readAnnotation(pod, CE_LABEL_PREFIX+EVENT_SOURCE_KEY)
+	errs = errs.Also(labelerr)
+
+	ceType, labelerr := readAnnotation(pod, CE_LABEL_PREFIX+EVENT_TYPE_KEY)
+	errs = errs.Also(labelerr)
+
+	// TODO(spencer-p) This is the only item not found in the pod - make sense in a field error?
+	img := os.Getenv(IMAGE_KEY)
+	if img == "" {
+		errs = errs.Also(apis.ErrMissingField("$" + IMAGE_KEY))
+	}
+
+	return &SidecarArgs{
+		SinkURIVar:      srcSinkURI,
+		OutputFormatVar: srcOutputFormat,
+		Image:           img,
+		Port:            port,
+		EventSource:     ceSrc,
+		EventType:       ceType,
+	}, errs
+}
+
+// injectSidecar rewrites the containers of the pod such that the sidecar is injected as
+// configured by SideCarArgs.
+func injectSidecar(pod *corev1.Pod, args *SidecarArgs) {
+	// Construct new container
+	portStr := strconv.Itoa(int(args.Port))
+	sidecarContainer := corev1.Container{
+		Name:  SIDECAR_NAME,
+		Image: args.Image,
+		Ports: []corev1.ContainerPort{{
+			ContainerPort: args.Port,
+		}},
+		Env: []corev1.EnvVar{{
+			Name:  "PORT",
+			Value: portStr,
+		}, {
+			Name:  "K_SINK",
+			Value: args.SinkURIVar.Value,
+		}, {
+			Name:  "K_OUTPUT_FORMAT",
+			Value: args.OutputFormatVar.Value,
+		}, {
+			Name:  "EVENT_SOURCE",
+			Value: args.EventSource,
+		}, {
+			Name:  "EVENT_TYPE",
+			Value: args.EventType,
+		}},
+	}
+
+	// Rewire the source container
+	args.SinkURIVar.Value = "http://127.0.0.1:" + portStr
+
+	// Add the sidecar container
+	pod.Spec.Containers = append(pod.Spec.Containers, sidecarContainer)
+}
+
+// Inject inspects and rewrites the Pod to have a Knative Adapter/Filter sidecar.
+func Inject(pod *corev1.Pod) {
+	args, errs := constructArgs(pod)
+	if errs != nil {
+		log.Printf("Cannot inject sidecar: %v\n", errs)
+		return
+	}
+
+	injectSidecar(pod, args)
+}
+
+// getSourceContainer finds a container that looks like it was configured as a source. It returns
+// the container itself, pointers to its useful environment variables, and an OK bool signifying
+// that the container was found.
+func getSourceContainer(pod *corev1.Pod) (
+	container *corev1.Container,
+	sinkURI *corev1.EnvVar,
+	outputFormat *corev1.EnvVar,
+	ok bool) {
+
+	for i := range pod.Spec.Containers {
+		for j, evar := range pod.Spec.Containers[i].Env {
+			switch evar.Name {
+			case "K_SINK":
+				sinkURI = &pod.Spec.Containers[i].Env[j]
+				container = &pod.Spec.Containers[i]
+			case "K_OUTPUT_FORMAT":
+				outputFormat = &pod.Spec.Containers[i].Env[j]
+				container = &pod.Spec.Containers[i]
+			}
+		}
+
+		// If we found the container, don't inspect others.
+		// This guarantees we will not take K_SINK from one container and K_OUTPUT_FORMAT from another.
+		if container != nil {
```

Comment:
I am not gonna say to solve this now, but you could have two containers that both want to use the sidecar right? is there a TODO that should be added to think about how a pod with two sources that are both emitting things but want to use the sidecar?

Reply:
Good point. These fns should all operate on slices of containers.